datafusion/physical-plan/src/joins/dynamic_filters.rs

```diff
@@ -0,0 +1,267 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::AsArray;
+use arrow::compute::filter_record_batch;
+use arrow::datatypes::DataType;
+use arrow::record_batch::RecordBatch;
+use datafusion_common::{exec_err, DataFusionError};
+use datafusion_expr::Accumulator;
+use datafusion_expr::Operator;
+use datafusion_functions_aggregate::min_max::{MaxAccumulator, MinAccumulator};
+use datafusion_physical_expr::expressions::{BinaryExpr, Column, Literal};
+use datafusion_physical_expr::PhysicalExpr;
+use hashbrown::HashSet;
+use parking_lot::Mutex;
+use std::fmt;
+use std::sync::Arc;
+
+pub struct DynamicFilterInfo {
+    columns: Vec<Arc<Column>>,
+    build_side_names: Vec<String>,
+    inner: Mutex<DynamicFilterInfoInner>,
+}
+
+struct DynamicFilterInfoInner {
+    max_accumulators: Vec<MaxAccumulator>,
+    min_accumulators: Vec<MinAccumulator>,
+    final_expr: Option<Arc<dyn PhysicalExpr>>,
+    batch_count: usize,
+    processed_partitions: HashSet<usize>,
+}
+
+impl fmt::Debug for DynamicFilterInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DynamicFilterInfo")
+            .field("columns", &self.columns)
+            .field("inner", &"<locked>")
+            .finish()
+    }
+}
+
+impl DynamicFilterInfo {
+    // Create a new DynamicFilterInfo instance
+    pub fn try_new(
+        columns: Vec<Arc<Column>>,
+        build_side_names: Vec<String>,
+        data_types: Vec<&DataType>,
+        total_batches: usize,
+    ) -> Result<Self, DataFusionError> {
+        let (max_accumulators, min_accumulators) = data_types
```

Comment:
I wonder if we can use a logical plan instead of manually invoking min/max accumulators? Not sure if that makes sense :)

Reply:
I don't know either, First I want to try using min/max ScalarFunction, but the ScalaFunction translation from logical_expr to physical_expr is not implemented yet... I guess using accumulator would be more Intuitive？